File: 范例/Accpt/18xx/p1802.pas
PROGRAM p1802;

VAR
        n,i,part                        :Integer;
        x,y                             :array[1..12]of Extended;
        whether,cover                   :array[1..12,1..12]of Boolean;
        down,le                         :array[1..12]of Integer;


FUNCTION Dis(a,b:Integer):Extended;
begin
dis:=sqrt(sqr(x[a]-x[b]) + sqr(y[a]-y[b]));
end;

PROCEDURE Main;
var
        i,j,col,k,max                  :Integer;
begin
fillchar(whether,sizeof(whether),0);
fillchar(down,sizeof(down),0);
fillchar(cover,sizeof(cover),0);
fillchar(le,sizeof(le),0);

for i:=1 to n do
        for j:=i+1 to n do
                if (dis(i,j)<20+1e-10) then
                        begin
                        whether[i][j]:=true;
                        whether[j][i]:=true;
                        end;

// make whether

col:=0;
for i:=1 to n do
        begin
        max:=-maxint;
        for j:=1 to n do
                if (le[j]=0) and (down[j]>max) then
                        begin
                        max:=down[j];
                        k:=j;
                        end;

        for j:=1 to n do if not cover[k][j] then break;
        le[k]:=j;
        if le[k]>col then col:=le[k];

        for j:=1 to n do
                if whether[k][j] and not cover[j][le[k]] then
                        begin
                        cover[j][le[k]]:=true;
                        inc(down[j]);
                        end;


        end;

writeln('The towers in case ',part,' can be covered in ',col,' frequencies.');

end;


BEGIN
//assign(input,'p.in');
//reset(input);

part:=0;
readln(n);
while (n<>0) do
        begin
        inc(part);
        for i:=1 to n do read(x[i],y[i]);
        readln;
        Main;
        readln(n);
        end;
END.

File: Vol.09(1800-1899)/#1802 Phone Home/p1802.cpp
// 1664877 2008-10-12 23:26:04 Accepted  1802 C++ 0 184 呆滞的慢板 

// 平面图着色，贪心就可以过，连回溯都没，可是算法正确性无法证明。 

#include <iostream>
#include <algorithm>
#include <cmath>
using namespace std;

double X[12], Y[12];
int N;
bool G[12][12], K[12];
int C[12], tot, mn;

bool near(int i, int j) {
    return sqrt((X[i]-X[j])*(X[i]-X[j]) +
        (Y[i]-Y[j])*(Y[i]-Y[j])) < 20.0000001;
}

void dfs(int v) {
    bool B[12] = {};
    for(int i = 0; i < N; ++i)
        if(G[v][i])
            B[C[i]] = true;
    for(int c = 1; c <= tot; ++c) {
        if(!B[c]) {
            C[v] = c;
            for(int i = 0; i < N; ++i)
                if(G[v][i] && !C[i])
                    dfs(i);
            return;
        }
    }
    C[v] = ++tot;
    for(int i = 0; i < N; ++i)
        if(G[v][i] && !C[i])
            dfs(i);
}

int main() {
    int T = 1;
    while(cin >> N && N) {
        for(int i = 0; i < N; ++i)
            cin >> X[i] >> Y[i];
        memset(G, 0, sizeof(G));
        for(int i = 0; i < N; ++i)
            for(int j = i + 1; j < N; ++j)
                G[i][j] = G[j][i] = near(i, j);
        tot = 0;
        memset(C, 0, sizeof(C));
        for(int i = 0; i < N; ++i)
            if(!C[i]) dfs(i);
        printf("The towers in case %d can be"
            " covered in %d frequencies.\n", T++, tot);
    }
}
